### ISSUE TITLE
DRBG fork safety and OPENSSL_INIT_ATFORK in 1.1.1d

### ISSUE THREAD

**Issue by reaperhulk** [openssl/openssl#9859]
With the change in 1.1.1d the built-in DRBG is now fork safe by default (🎉). Given that, does `OPENSSL_INIT_ATFORK` still have a purpose? It still registers the fork handlers, but they no longer do anything. Should this be documented in some fashion?

**Comment by mspncp** [comment]
Personally, I would have liked to keep the fork handler approach, because it is simple. Unfortunately, the fork handlers were disabled by default some time between the initial implementation and the first release of 1.1.1, and this cannot be changed anymore in a stable release. Also, some people are reluctant to let OpenSSL install fork handlers by default. OTHOH, the current getpid() protection comes at the price of a lot of getpid() calls, see the discussion with @kroeckx in https://github.com/openssl/openssl/pull/9802#issuecomment-529205999 and following.

I have some ideas for improvements, but it was too late to add them to the current release:

- Reduce the number of getpid() calls
- Restore the fork handler fork-protection and use it when libcrypto is initialized with the `OPENSSL_INIT_ATFORK` flag. Otherwise, use the getpid() protection. 

In other words: `OPENSSL_INIT_ATFORK` is not buried yet.

**Comment by reaperhulk** [comment]
@mspncp Okay -- so INIT_ATFORK may turn into an optimization path to avoid the large quantity of `getpid` calls that result from this new approach, but consumers can safely assume that OpenSSL is fork safe starting with 1.1.1d moving forward?

**Comment by mspncp** [comment]
Yes. Fork-protection was our intention for 1.1.1 from the very beginning and it was unintentionally disabled by default. 

Note also that I added a testcase to check the reseeding after fork, so this accident should not happen again.

**Comment by mspncp** [comment]
_(@reaperhulk please don't wonder about my last post, which I deleted immediately. I got interrupted by a WhatsApp message and replied on the wrong forum :blush:)_

**Comment by reaperhulk** [comment]
Okay, thanks for the clarification!
